<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>kmerExtractor &mdash; kmerExtractor 1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            kmerExtractor
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../kmerExtractorTutorial.html">kmerExtractor user manual</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">kmerExtractor</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Module code</a></li>
      <li class="breadcrumb-item active">kmerExtractor</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for kmerExtractor</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">itertools</span> <span class="kn">import</span> <span class="n">product</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">Counter</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<div class="viewcode-block" id="kmers_in_sequence"><a class="viewcode-back" href="../docs_prev/kmerExtractor.html#kmerExtractor.kmers_in_sequence">[docs]</a><span class="k">def</span> <span class="nf">kmers_in_sequence</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span><span class="n">k</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Calculate the frequency of k-mers in a given sequence.</span>
<span class="sd">    Only k-mers containing the four basic uppercase nucleotides (A, C, G, T) are accepted.</span>
<span class="sd">    Any substrings of length k that have any other character in their composition will be disregarded and thus not considered as k-mers.</span>

<span class="sd">    :param sequence: Nucleotide sequence.</span>
<span class="sd">    :type sequence: str</span>
<span class="sd">    :param k: Length of k-mers.</span>
<span class="sd">    :type k: int</span>

<span class="sd">    :return: Dictionary containing the k-mer frequencies. Keys are the k-mers and values are the corresponding frequencies.</span>
<span class="sd">    :rtype: dict</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">kmers_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">km</span><span class="p">)</span> <span class="k">for</span> <span class="n">km</span> <span class="ow">in</span> <span class="n">product</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">,</span> <span class="n">repeat</span> <span class="o">=</span> <span class="n">k</span><span class="p">)]</span>
    <span class="n">kmers</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">kmers_list</span><span class="p">}</span>
    <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="k">for</span> <span class="n">nt</span> <span class="ow">in</span> <span class="n">sequence</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">key</span><span class="p">)</span><span class="o">==</span><span class="n">k</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">kmers</span><span class="p">:</span>
                <span class="n">kmers</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">key</span><span class="o">=</span><span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span><span class="o">+</span><span class="n">nt</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">key</span> <span class="o">=</span> <span class="n">key</span> <span class="o">+</span> <span class="n">nt</span>
    <span class="c1"># count last k-mer</span>
    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">key</span><span class="p">)</span> <span class="o">==</span> <span class="n">k</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">key</span> <span class="ow">in</span> <span class="n">kmers</span><span class="p">):</span>
        <span class="n">kmers</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">kmers</span></div>


<div class="viewcode-block" id="kmers_by_window"><a class="viewcode-back" href="../docs_prev/kmerExtractor.html#kmerExtractor.kmers_by_window">[docs]</a><span class="k">def</span> <span class="nf">kmers_by_window</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">window_size</span><span class="o">=</span><span class="mi">100_000</span><span class="p">,</span> <span class="n">output_path</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Calculate the frequency of k-mers by windows within each of the chromosome sequences.</span>
<span class="sd">    First, the entire sequence contained within the FASTA file is read.</span>
<span class="sd">    The sequences of each chromosome are then identified and separated.</span>
<span class="sd">    Subsequently, each chromosome is separated into windows of the specified size.</span>
<span class="sd">    Finally, the frequencies of the k-mers are calculated for each window.</span>

<span class="sd">    :param filepath: Location of the FASTA file containing the nucleotide sequence of an organism separated by chromosomes.</span>
<span class="sd">    :type filepath: str</span>
<span class="sd">    :param k: Length of k-mers.</span>
<span class="sd">    :type k: int</span>
<span class="sd">    :param window_size: Window size, in number of base pairs, to be used to count the frequency of k-mers within the chromosome.</span>
<span class="sd">                        Must be less than or equal to the length of the smallest chromosome.</span>
<span class="sd">                        Default to 100000.</span>
<span class="sd">    :type window_size: int</span>
<span class="sd">    :param output_path: Path to the directory where the output csv file will be saved.</span>
<span class="sd">                        If not specified, the output file will be saved in the same directory as the input file. Default to None</span>
<span class="sd">    :type output_path: str</span>

<span class="sd">    :return: Dataframe indicating the chromosome, the window index, the start and end positions of the window,</span>
<span class="sd">             and the frequencies of the corresponding k-mers.</span>
<span class="sd">    :rtype: pandas.DataFrame</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="k">def</span> <span class="nf">hop</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">step</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">step</span><span class="p">):</span>
            <span class="k">yield</span> <span class="n">i</span>
        <span class="k">yield</span> <span class="n">stop</span>

    <span class="n">f_in</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
    <span class="n">all_file</span> <span class="o">=</span> <span class="n">f_in</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
    <span class="n">f_in</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="n">path</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>
    <span class="n">file_noext</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">filename</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">output_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">outputfile_path</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">+</span> <span class="s1">&#39;/CGRW_k</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">.csv&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">file_noext</span><span class="p">)</span>  <span class="c1"># save in the specified directory</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">outputfile_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">+</span> <span class="s1">&#39;/CGRW_k</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">.csv&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">file_noext</span><span class="p">)</span>  <span class="c1"># save in the input directory</span>

    <span class="n">f_out</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">outputfile_path</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span>

    <span class="n">kmers_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">km</span><span class="p">)</span> <span class="k">for</span> <span class="n">km</span> <span class="ow">in</span> <span class="n">product</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">,</span> <span class="n">repeat</span><span class="o">=</span><span class="n">k</span><span class="p">)]</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;chromosome,window,start,end&#39;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">f_out</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="o">*</span><span class="n">kmers_list</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">f_out</span><span class="p">)</span>

    <span class="n">seq_all</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">all_file</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">))</span>
    <span class="k">del</span> <span class="n">all_file</span>
    <span class="n">seq_bychr</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;&gt;chr[1-9]*&#39;</span><span class="p">,</span> <span class="n">seq_all</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]</span>
    <span class="n">chr_names</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="s1">&#39;&gt;chr[1-9]*&#39;</span><span class="p">,</span> <span class="n">seq_all</span><span class="p">)</span>
    <span class="n">chr_len</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">seq_bychr</span><span class="p">]</span>
    <span class="k">del</span> <span class="n">seq_all</span>

    <span class="k">if</span> <span class="n">window_size</span> <span class="o">&gt;</span> <span class="nb">min</span><span class="p">(</span><span class="n">chr_len</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Window size must be less than or equal to the length of the smallest chromosome.&#39;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">chr_ix</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq_bychr</span><span class="p">)):</span>
        <span class="n">position</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">hop</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_bychr</span><span class="p">[</span><span class="n">chr_ix</span><span class="p">]),</span> <span class="n">window_size</span><span class="p">))</span>
        <span class="n">window</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">position</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">position</span><span class="p">[</span><span class="mi">1</span><span class="p">:]):</span>
            <span class="n">chr_seq</span> <span class="o">=</span> <span class="n">seq_bychr</span><span class="p">[</span><span class="n">chr_ix</span><span class="p">][</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span>
            <span class="n">kmers_w</span> <span class="o">=</span> <span class="n">kmers_in_sequence</span><span class="p">(</span><span class="n">chr_seq</span><span class="p">,</span> <span class="n">k</span><span class="p">)</span>

            <span class="nb">print</span><span class="p">(</span><span class="n">chr_names</span><span class="p">[</span><span class="n">chr_ix</span><span class="p">][</span><span class="mi">1</span><span class="p">:]</span> <span class="o">+</span> <span class="s1">&#39;,&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">window</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;,&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">start</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;,&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">end</span><span class="p">),</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span>
                  <span class="n">file</span><span class="o">=</span><span class="n">f_out</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="o">*</span><span class="nb">list</span><span class="p">(</span><span class="n">kmers_w</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">f_out</span><span class="p">)</span>

            <span class="n">window</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="n">f_out</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">outputfile_path</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">df</span></div>


<div class="viewcode-block" id="kmers_by_window_opt"><a class="viewcode-back" href="../docs_prev/kmerExtractor.html#kmerExtractor.kmers_by_window_opt">[docs]</a><span class="k">def</span> <span class="nf">kmers_by_window_opt</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">window_size</span><span class="o">=</span><span class="mi">100_000</span><span class="p">,</span> <span class="n">output_path</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Calculate the frequency of k-mers by windows within each of the chromosome sequences.</span>
<span class="sd">    It calculates the k-mers, as it reads the FASTA file, optimizing space by avoiding storing the sequence.</span>

<span class="sd">    :param filepath: Location of the FASTA file containing the nucleotide sequence of an organism separated by chromosomes.</span>
<span class="sd">    :type filepath: str</span>
<span class="sd">    :param k: Length of k-mers.</span>
<span class="sd">    :type k: int</span>
<span class="sd">    :param window_size: Window size, in number of base pairs, to be used to count the frequency of k-mers within the chromosome.</span>
<span class="sd">                        Must be greater than the number of nucleotides per line in the input FASTA file. Default to 100000</span>
<span class="sd">    :type window_size: int</span>
<span class="sd">    :param output_path: Path to the directory where the output csv file will be saved.</span>
<span class="sd">                        If not specified, the output file will be saved in the same directory as the input file. Default to None</span>
<span class="sd">    :type output_path: str</span>

<span class="sd">    :return: Dataframe containing the k-mer frequencies for all windows of each chromosome.</span>
<span class="sd">    :rtype: pandas.DataFrame</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="n">kmers_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">km</span><span class="p">)</span> <span class="k">for</span> <span class="n">km</span> <span class="ow">in</span> <span class="n">product</span><span class="p">(</span><span class="s1">&#39;ACGT&#39;</span><span class="p">,</span> <span class="n">repeat</span><span class="o">=</span><span class="n">k</span><span class="p">)]</span>

    <span class="n">input_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
    <span class="n">path</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>
    <span class="n">file_noext</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">filename</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">output_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">outputfile_path</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">+</span> <span class="s1">&#39;/CGRW_k</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">.csv&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">file_noext</span><span class="p">)</span>  <span class="c1"># save in the specified directory</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">outputfile_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">+</span> <span class="s1">&#39;/CGRW_k</span><span class="si">{0}</span><span class="s1">_</span><span class="si">{1}</span><span class="s1">.csv&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">file_noext</span><span class="p">)</span>  <span class="c1"># save in the input directory</span>
    <span class="n">f</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">outputfile_path</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span>

    <span class="n">kmers</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="p">:</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">kmers_list</span><span class="p">}</span>
    <span class="c1">#key = &#39;&#39;</span>
    <span class="n">nts_in_line_prev</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;chromosome,window,start,end&#39;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">f</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="o">*</span><span class="nb">list</span><span class="p">(</span><span class="n">kmers</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">f</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">input_file</span><span class="o">.</span><span class="n">readlines</span><span class="p">()):</span>
        <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;&gt;&#39;</span><span class="p">:</span>  <span class="c1"># new chromosome start</span>
            <span class="n">window</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">start</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">end</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">chromosome</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="c1"># the first and last characters are &#39;&gt;&#39; and &#39;\n&#39;, respectively</span>
            <span class="n">kmers</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="p">:</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">kmers_list</span><span class="p">}</span>
            <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>

        <span class="k">else</span><span class="p">:</span>  <span class="c1"># if line[0] != &#39;&gt;&#39;:</span>
            <span class="n">nts_in_line</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>  <span class="c1"># last character is &#39;\n&#39;</span>
            <span class="n">count</span> <span class="o">+=</span> <span class="n">nts_in_line</span>

            <span class="k">if</span> <span class="n">nts_in_line_prev</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">nts_in_line_prev</span> <span class="o">=</span> <span class="n">nts_in_line</span>

            <span class="k">if</span> <span class="n">count</span> <span class="o">&lt;=</span> <span class="n">window_size</span><span class="p">:</span>
                <span class="n">stop</span> <span class="o">=</span> <span class="n">nts_in_line</span>  <span class="c1"># count k-mers until line end</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">stop</span> <span class="o">=</span> <span class="n">nts_in_line</span> <span class="o">-</span> <span class="p">(</span><span class="n">count</span> <span class="o">-</span> <span class="n">window_size</span><span class="p">)</span>  <span class="c1"># count k-mers until window end</span>
                <span class="n">count</span> <span class="o">-=</span> <span class="n">nts_in_line</span> <span class="o">-</span> <span class="n">stop</span>

            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">stop</span><span class="p">):</span>
                <span class="n">nt</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                <span class="n">end</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">key</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">k</span><span class="p">:</span>
                    <span class="n">key</span> <span class="o">=</span> <span class="n">key</span> <span class="o">+</span> <span class="n">nt</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">kmers</span><span class="p">:</span>
                        <span class="n">kmers</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">key</span> <span class="o">=</span> <span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="o">+</span> <span class="n">nt</span>

            <span class="k">if</span> <span class="n">count</span> <span class="o">&gt;=</span> <span class="n">window_size</span><span class="p">:</span>  <span class="c1"># window is over</span>
                <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">kmers</span><span class="p">:</span>
                    <span class="n">kmers</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>  <span class="c1"># add last k-mer in window</span>
                <span class="c1"># print window info and k-mer frequencies</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">window</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="o">*</span><span class="nb">list</span><span class="p">(</span><span class="n">kmers</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">f</span><span class="p">)</span>

                <span class="c1"># new window info</span>
                <span class="n">window</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">start</span> <span class="o">=</span> <span class="n">end</span> <span class="o">+</span> <span class="mi">1</span>
                <span class="n">end</span> <span class="o">=</span> <span class="n">start</span> <span class="o">-</span> <span class="mi">1</span>
                <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">kmers</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="p">:</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">kmers_list</span><span class="p">}</span>
                <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>

                <span class="c1"># start new window counting in the rest of the line</span>
                <span class="n">count</span> <span class="o">+=</span> <span class="n">nts_in_line</span> <span class="o">-</span> <span class="n">stop</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">stop</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
                    <span class="n">nt</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                    <span class="n">end</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">key</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">k</span><span class="p">:</span>
                        <span class="n">key</span> <span class="o">=</span> <span class="n">key</span> <span class="o">+</span> <span class="n">nt</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">kmers</span><span class="p">:</span>
                            <span class="n">kmers</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">key</span> <span class="o">=</span> <span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="o">+</span> <span class="n">nt</span>

            <span class="k">if</span> <span class="n">nts_in_line</span> <span class="o">&lt;</span> <span class="n">nts_in_line_prev</span><span class="p">:</span>  <span class="c1"># chromosome is over but window is not</span>
                <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">kmers</span><span class="p">:</span>
                    <span class="n">kmers</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>  <span class="c1"># add last k-mer in window</span>
                <span class="c1"># print window info and k-mer frequencies</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">window</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="o">*</span><span class="nb">list</span><span class="p">(</span><span class="n">kmers</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">f</span><span class="p">)</span>

            <span class="n">nts_in_line_prev</span> <span class="o">=</span> <span class="n">nts_in_line</span>

    <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">kmers</span><span class="p">:</span>
        <span class="n">kmers</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>  <span class="c1"># last k-mer in window</span>
    <span class="c1"># print window info and k-mer frequencies</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">window</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="o">*</span><span class="nb">list</span><span class="p">(</span><span class="n">kmers</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="n">f</span><span class="p">)</span>

    <span class="n">f</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">outputfile_path</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">df</span></div>


<div class="viewcode-block" id="FCGR"><a class="viewcode-back" href="../docs_prev/kmerExtractor.html#kmerExtractor.FCGR">[docs]</a><span class="k">def</span> <span class="nf">FCGR</span><span class="p">(</span><span class="n">kmers</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Organize the k-mers and their associated values in the common structure of </span>
<span class="sd">    Frequency Chaos Game Representation (FCGR), i.e., a matrix of dimensions N x N,</span>
<span class="sd">    where N = 4^(k/2); with the upper left corner equal to the k-mer of C, </span>
<span class="sd">    the upper right corner the k-mer of G, the lower left corner the k-mer of A,</span>
<span class="sd">    and the lower right corner the k-mer of T.</span>

<span class="sd">    :param kmers: dictionary mapping the k-mers to a real value. All k-mers must be of the same length.</span>
<span class="sd">    :type kmers: dict</span>

<span class="sd">    :returns: FCGR matrix and the names of the k-mers in the matrix.</span>
<span class="sd">    :rtype: tuple[numpy.ndarray, numpy.ndarray]</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">k</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">kmers</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">N</span> <span class="o">=</span> <span class="mi">2</span><span class="o">**</span><span class="n">k</span>
    <span class="n">counts</span> <span class="o">=</span> <span class="p">[[</span><span class="mi">0</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">)]</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">)]</span>
    <span class="n">names</span> <span class="o">=</span> <span class="p">[[</span><span class="kc">None</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">)]</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N</span><span class="p">)]</span>
    <span class="n">maxx</span><span class="p">,</span> <span class="n">maxy</span> <span class="o">=</span> <span class="n">N</span><span class="p">,</span> <span class="n">N</span>
    <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span><span class="n">value</span> <span class="ow">in</span> <span class="n">kmers</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">rkey</span> <span class="o">=</span> <span class="n">key</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">char</span> <span class="ow">in</span> <span class="n">rkey</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">char</span> <span class="o">==</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span> <span class="n">x</span> <span class="o">+=</span> <span class="n">maxx</span> <span class="o">&gt;&gt;</span> <span class="mi">1</span>     <span class="c1"># maxx // 2</span>
            <span class="k">elif</span> <span class="n">char</span> <span class="o">==</span> <span class="s1">&#39;A&#39;</span><span class="p">:</span> <span class="n">y</span> <span class="o">+=</span> <span class="n">maxy</span> <span class="o">&gt;&gt;</span> <span class="mi">1</span>  <span class="c1"># maxy // 2</span>
            <span class="k">elif</span> <span class="n">char</span> <span class="o">==</span> <span class="s1">&#39;T&#39;</span> <span class="ow">or</span> <span class="n">char</span> <span class="o">==</span> <span class="s1">&#39;U&#39;</span><span class="p">:</span> 
                <span class="n">x</span> <span class="o">+=</span> <span class="n">maxx</span> <span class="o">&gt;&gt;</span> <span class="mi">1</span>
                <span class="n">y</span> <span class="o">+=</span> <span class="n">maxy</span> <span class="o">&gt;&gt;</span> <span class="mi">1</span>
            <span class="n">maxx</span> <span class="o">&gt;&gt;=</span> <span class="mi">1</span>
            <span class="n">maxy</span> <span class="o">&gt;&gt;=</span> <span class="mi">1</span>
        
        <span class="n">counts</span><span class="p">[</span><span class="n">y</span><span class="p">][</span><span class="n">x</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>  
        <span class="n">names</span><span class="p">[</span><span class="n">y</span><span class="p">][</span><span class="n">x</span><span class="p">]</span> <span class="o">=</span> <span class="n">key</span>
        <span class="n">maxx</span><span class="p">,</span> <span class="n">maxy</span> <span class="o">=</span> <span class="n">N</span><span class="p">,</span> <span class="n">N</span>
        <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span>
 
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">counts</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">names</span><span class="p">)</span></div>


<div class="viewcode-block" id="plot_kmers_across_windows"><a class="viewcode-back" href="../docs_prev/kmerExtractor.html#kmerExtractor.plot_kmers_across_windows">[docs]</a><span class="k">def</span> <span class="nf">plot_kmers_across_windows</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">kmer_names</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Plot the frequencies of the given k-mers across the windows of a given chromosome.</span>

<span class="sd">    :param df: dataframe with the k-mer frequencies by window.</span>
<span class="sd">    :type df: pandas.DataFrame</span>
<span class="sd">    :param kmer_names: list of k-mer names to plot.</span>
<span class="sd">    :type kmer_names: list</span>
<span class="sd">    :param chromosome: name of the chromosome to plot.</span>
<span class="sd">    :type chromosome: str</span>
<span class="sd">    :param figsize: The size in inches of the figure to create. default to (12,4).</span>
<span class="sd">    :type figsize: tuple[int,int]</span>
<span class="sd">    :param ax: The Axes object containing the plot. If None, a new figure and axes is created. Default to None.</span>
<span class="sd">    :type ax: matplotlib.axes._axes.Axes</span>

<span class="sd">    :return: The Axes object containing the plot.</span>
<span class="sd">    :rtype: matplotlib.axes._axes.Axes</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">window_size</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">fig</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">ax</span>

    <span class="k">for</span> <span class="n">kmer</span> <span class="ow">in</span> <span class="n">kmer_names</span><span class="p">:</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">lineplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">chromosome</span> <span class="o">==</span> <span class="n">chromosome</span><span class="p">],</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;window&#39;</span><span class="p">,</span><span class="n">y</span><span class="o">=</span><span class="n">kmer</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">kmer</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Chromosome </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">chromosome</span><span class="p">),</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Chromosome position (x</span><span class="si">{}</span><span class="s1"> bp)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">window_size</span><span class="p">),</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;k-mer frequency&#39;</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">ax</span></div>


<div class="viewcode-block" id="plot_kmers_freq_within_chromosomes"><a class="viewcode-back" href="../docs_prev/kmerExtractor.html#kmerExtractor.plot_kmers_freq_within_chromosomes">[docs]</a><span class="k">def</span> <span class="nf">plot_kmers_freq_within_chromosomes</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span><span class="mi">6</span><span class="p">),</span><span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Plot the sum of the k-mer frequencies within each chromosome.</span>

<span class="sd">    :param df: dataframe with the k-mer frequencies by window.</span>
<span class="sd">    :type df: pandas.DataFrame</span>
<span class="sd">    :param figsize: The size in inches of the figure to create. default to (15,6).</span>
<span class="sd">    :type figsize: tuple[int,int]</span>
<span class="sd">    :param ax: Axes in which to draw the plot, otherwise use the currently-active Axes. Default to None.</span>
<span class="sd">    :type ax: matplotlib.axes.Axes</span>

<span class="sd">    :return: The Axes object containing the plot.</span>
<span class="sd">    :rtype: matplotlib.axes._axes.Axes</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">df_agg</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;chromosome&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span><span class="n">k</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">4</span><span class="p">:]})</span><span class="o">.</span><span class="n">T</span>

    <span class="n">axi</span> <span class="o">=</span> <span class="n">df_agg</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kind</span><span class="o">=</span><span class="s1">&#39;bar&#39;</span><span class="p">,</span> <span class="n">stacked</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>

    <span class="n">axi</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;k-mer frequency&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
    <span class="n">axi</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">axi</span></div>


<div class="viewcode-block" id="plot_FCGR"><a class="viewcode-back" href="../docs_prev/kmerExtractor.html#kmerExtractor.plot_FCGR">[docs]</a><span class="k">def</span> <span class="nf">plot_FCGR</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">chromosome</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span> <span class="n">colormap</span><span class="o">=</span><span class="s1">&#39;bwr&#39;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Plot the Frequency Chaos Game Representation (FCGR) of the k-mers in the input dataframe.</span>

<span class="sd">    :param df: Dataframe with the k-mer frequencies by windows across chromosomes.</span>
<span class="sd">    :type df: pandas.DataFrame</span>
<span class="sd">    :param chromosome: Name of the chromosome to plot. If None, parameter window must be None,</span>
<span class="sd">                       plotting the sum of k-mers across all chromosomes. Default to None.</span>
<span class="sd">    :type chromosome: str</span>
<span class="sd">    :param window: Number of the window to plot. If None, plot the sum of all windows for each chromosome. Default to None.</span>
<span class="sd">    :type window: int</span>
<span class="sd">    :param figsize: The size in inches of the figure to create. Default to (7,6).</span>
<span class="sd">    :type figsize: tuple[int,int]</span>
<span class="sd">    :param colormap: Matplotlib colormap name. The mapping from data values to color space. Default to &#39;bwr&#39;.</span>
<span class="sd">    :type colormap: str</span>
<span class="sd">    :param ax: Axes in which to draw the plot, otherwise use the currently-active Axes. Default to None.</span>
<span class="sd">    :type ax: matplotlib.axes.Axes</span>

<span class="sd">    :return: The Axes object containing the plot.</span>
<span class="sd">    :rtype: matplotlib.axes._axes.Axes</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">kmer_names</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">4</span><span class="p">:]</span>
    <span class="k">if</span> <span class="n">window</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;chromosome&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span><span class="n">k</span><span class="p">:</span> <span class="s1">&#39;sum&#39;</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">4</span><span class="p">:]})</span><span class="o">.</span><span class="n">T</span>
        <span class="k">if</span> <span class="n">chromosome</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">vals</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">vals</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">chromosome</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">chromosome</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Parameter chromosome must be specified if window is not None.&#39;</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[(</span><span class="n">df</span><span class="o">.</span><span class="n">chromosome</span> <span class="o">==</span> <span class="n">chromosome</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">window</span> <span class="o">==</span> <span class="n">window</span><span class="p">)]</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">kmer_names</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="n">v_min</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">vals</span><span class="p">)</span>
    <span class="n">v_max</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">vals</span><span class="p">)</span>
    <span class="n">vals_norm</span> <span class="o">=</span> <span class="p">(</span><span class="n">vals</span> <span class="o">-</span> <span class="n">v_min</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">v_max</span> <span class="o">-</span> <span class="n">v_min</span> <span class="o">+</span> <span class="mf">1e-6</span><span class="p">)</span>

    <span class="c1"># dictionary with normalized values</span>
    <span class="n">kmer2freq</span> <span class="o">=</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">kmer_names</span><span class="p">,</span> <span class="n">vals_norm</span><span class="p">)}</span>

    <span class="c1"># FCGR matrices</span>
    <span class="n">FCGR_vals</span><span class="p">,</span> <span class="n">FCGR_names</span> <span class="o">=</span> <span class="n">FCGR</span><span class="p">(</span><span class="n">kmer2freq</span><span class="p">)</span>

    <span class="c1"># -------------------------------------------</span>
    <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">ax</span>

    <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">FCGR_vals</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="n">FCGR_names</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
                <span class="n">yticklabels</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="n">xticklabels</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                <span class="n">square</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">cmap</span><span class="o">=</span><span class="n">colormap</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ax</span></div>


<div class="viewcode-block" id="CGR"><a class="viewcode-back" href="../docs_prev/kmerExtractor.html#kmerExtractor.CGR">[docs]</a><span class="k">def</span> <span class="nf">CGR</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">d</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Generate the X and Y coordinates of the Chaos Game Representation (CGR) for the input sequence.</span>
<span class="sd">    Purine nucleotides (A and G) are placed on the minor diagonal, and the pyrimidine nucleotides (C and T) on the main diagonal.</span>
<span class="sd">    The CGR is built in a square of dimensions 2d x 2d with center at the coordinates (0,0).</span>
<span class="sd">    Only the four basic uppercase nucleotides (A, C, G, T) are accepted. Any other character is ignored.</span>

<span class="sd">    :param sequence: Nucleotide sequence.</span>
<span class="sd">    :type sequence: str</span>
<span class="sd">    :param d: Half of the side of the square. Default to 1.</span>
<span class="sd">    :type d: int</span>

<span class="sd">    :return: The x and y coordinates of the CGR.</span>
<span class="sd">    :rtype: tuple[list[float],list[float]]</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">ACGT</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;A&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span> <span class="o">-</span> <span class="n">d</span><span class="p">,</span> <span class="mi">0</span> <span class="o">-</span> <span class="n">d</span><span class="p">],</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span> <span class="o">-</span> <span class="n">d</span><span class="p">,</span> <span class="mi">0</span> <span class="o">+</span> <span class="n">d</span><span class="p">],</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span> <span class="o">+</span> <span class="n">d</span><span class="p">,</span> <span class="mi">0</span> <span class="o">+</span> <span class="n">d</span><span class="p">],</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span> <span class="o">+</span> <span class="n">d</span><span class="p">,</span> <span class="mi">0</span> <span class="o">-</span> <span class="n">d</span><span class="p">]}</span>

    <span class="n">X</span><span class="p">,</span> <span class="n">Y</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">j</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">))):</span>
        <span class="n">nt</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">nt</span> <span class="ow">in</span> <span class="s1">&#39;ACGT&#39;</span><span class="p">:</span>
            <span class="n">x</span> <span class="o">=</span> <span class="p">(</span><span class="n">X</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">+</span> <span class="n">ACGT</span><span class="p">[</span><span class="n">nt</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span>
            <span class="n">y</span> <span class="o">=</span> <span class="p">(</span><span class="n">Y</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">+</span> <span class="n">ACGT</span><span class="p">[</span><span class="n">nt</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span>
            <span class="n">X</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
            <span class="n">Y</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
            <span class="n">j</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">return</span> <span class="n">X</span><span class="p">,</span> <span class="n">Y</span></div>


<div class="viewcode-block" id="plot_CGR"><a class="viewcode-back" href="../docs_prev/kmerExtractor.html#kmerExtractor.plot_CGR">[docs]</a><span class="k">def</span> <span class="nf">plot_CGR</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">d</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">)):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Plot the Chaos Game Representation (CGR) of the input sequence.</span>
<span class="sd">    Purine nucleotides (A and G) are placed on the minor diagonal, and the pyrimidine nucleotides (C and T) on the main diagonal.</span>
<span class="sd">    The CGR is built in a square of dimensions 2d x 2d with center at the coordinates (0,0).</span>
<span class="sd">    Only the four basic uppercase nucleotides (A, C, G, T) are accepted. Any other character is ignored.</span>

<span class="sd">    :param sequence: Nucleotide sequence.</span>
<span class="sd">    :type sequence: str</span>
<span class="sd">    :param d: Half of the side of the square in which the CGR is built. Default to 1.</span>
<span class="sd">    :type d: int</span>

<span class="sd">    :return: The Axes object containing the plot.</span>
<span class="sd">    :rtype: matplotlib.axes._axes.Axes</span>
<span class="sd">    &#39;&#39;&#39;</span>

    <span class="n">X</span><span class="p">,</span><span class="n">Y</span> <span class="o">=</span> <span class="n">CGR</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">d</span><span class="p">)</span>
    <span class="c1"># -------------------------------------------</span>
    <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="n">ax</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">Y</span><span class="p">,</span> <span class="s1">&#39;ko&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="n">markersize</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">((</span><span class="o">-</span><span class="n">d</span><span class="p">,</span> <span class="n">d</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">((</span><span class="o">-</span><span class="n">d</span><span class="p">,</span> <span class="n">d</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>

    <span class="k">return</span> <span class="n">ax</span></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Camila Riccio-Rengifo.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>