<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>kmerExtractor user manual &mdash; kmerExtractor 1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="kmerExtractor module" href="kmerExtractor.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            kmerExtractor
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="modules.html">Documentation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">kmerExtractor user manual</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Import-libraries">Import libraries</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Compute-k-mers">Compute k-mers</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Read-k-mer-files-if-already-computed">Read k-mer files if already computed</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Group-all-windows-within-each-chromosome">Group all windows within each chromosome</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Group-all-chromosomes">Group all chromosomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="#k-mer-figures">k-mer figures</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Plot-k-mers-across-windows">Plot k-mers across windows</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Plot-k-mer-frequencies-within-each-chromosome">Plot k-mer frequencies within each chromosome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Plot-FCGR-structure">Plot FCGR structure</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Plot-CGR">Plot CGR</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">kmerExtractor</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">kmerExtractor user manual</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/kmerExtractorTutorial.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="kmerExtractor-user-manual">
<h1>kmerExtractor user manual<a class="headerlink" href="#kmerExtractor-user-manual" title="Permalink to this headline"></a></h1>
<p>Below we present examples of how to use the main functions of the software.</p>
<section id="Import-libraries">
<h2>Import libraries<a class="headerlink" href="#Import-libraries" title="Permalink to this headline"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="c1"># software to extract the k-mers</span>
<span class="kn">import</span> <span class="nn">kmerExtractor</span> <span class="k">as</span> <span class="nn">kex</span>
</pre></div>
</div>
</div>
</section>
<section id="Compute-k-mers">
<h2>Compute k-mers<a class="headerlink" href="#Compute-k-mers" title="Permalink to this headline"></a></h2>
<p>Specify the corresponding directory of the FASTA files from which you want to extract the k-mers, and the directory where you want to save the output files. Also define the window size (in base pairs) for which you want the frequency of k-mers to be counted.</p>
<p>In this case we will use the genome of arabidospsis, tomato and maize, with a window size of <span class="math notranslate nohighlight">\(100000\)</span> bp for all genomes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">filepath_ara</span> <span class="o">=</span> <span class="s1">&#39;../input_data/arabidopsis_genome.fasta&#39;</span>
<span class="n">filepath_tom</span> <span class="o">=</span> <span class="s1">&#39;../input_data/tomato_genome.fasta&#39;</span>
<span class="n">filepath_mz</span> <span class="o">=</span> <span class="s1">&#39;../input_data/maize_genome.fasta&#39;</span>
<span class="n">output_path</span> <span class="o">=</span> <span class="s1">&#39;../output_data&#39;</span>
<span class="n">window_size</span> <span class="o">=</span> <span class="mi">100_000</span>
</pre></div>
</div>
</div>
<p>Now use the <code class="docutils literal notranslate"><span class="pre">kmers_by_window_opt</span></code> function to calculate the k-mers of each genome for the <span class="math notranslate nohighlight">\(k\)</span> of your choice.</p>
<p>In the output file, the first column indicates the chromosome, the second column the window index, the third and fourth columns indicate the start and end position of the window respectively, and the rest of the <span class="math notranslate nohighlight">\(4^k\)</span> columns correspond to the k-mer frequencies. This data is saved in a csv file and stored in a dataframe for downstream analysis within the tutorial.</p>
<p>The resulting dataframe for arabidopsis 2-mers looks like shown below.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_2mers_ara</span> <span class="o">=</span> <span class="n">kex</span><span class="o">.</span><span class="n">kmers_by_window_opt</span><span class="p">(</span><span class="n">filepath_ara</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">window_size</span><span class="o">=</span><span class="n">window_size</span><span class="p">,</span> <span class="n">output_path</span><span class="o">=</span><span class="n">output_path</span><span class="p">)</span>
<span class="n">df_2mers_tom</span> <span class="o">=</span> <span class="n">kex</span><span class="o">.</span><span class="n">kmers_by_window_opt</span><span class="p">(</span><span class="n">filepath_tom</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">window_size</span><span class="o">=</span><span class="n">window_size</span><span class="p">,</span> <span class="n">output_path</span><span class="o">=</span><span class="n">output_path</span><span class="p">)</span>
<span class="n">df_2mers_mz</span>  <span class="o">=</span> <span class="n">kex</span><span class="o">.</span><span class="n">kmers_by_window_opt</span><span class="p">(</span><span class="n">filepath_mz</span><span class="p">,</span>  <span class="n">k</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">window_size</span><span class="o">=</span><span class="n">window_size</span><span class="p">,</span> <span class="n">output_path</span><span class="o">=</span><span class="n">output_path</span><span class="p">)</span>
<span class="n">df_2mers_ara</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|████████████████████████████████████████| 1537509/1537509 [00:45&lt;00:00, 33733.02it/s]
100%|██████████████████████████████████████| 10090326/10090326 [04:45&lt;00:00, 35388.84it/s]
100%|██████████████████████████████████████| 25502007/25502007 [12:31&lt;00:00, 33940.82it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chromosome</th>
      <th>window</th>
      <th>start</th>
      <th>end</th>
      <th>AA</th>
      <th>AC</th>
      <th>AG</th>
      <th>AT</th>
      <th>CA</th>
      <th>CC</th>
      <th>CG</th>
      <th>CT</th>
      <th>GA</th>
      <th>GC</th>
      <th>GG</th>
      <th>GT</th>
      <th>TA</th>
      <th>TC</th>
      <th>TG</th>
      <th>TT</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>0</td>
      <td>1</td>
      <td>100000</td>
      <td>11617</td>
      <td>5201</td>
      <td>5956</td>
      <td>9480</td>
      <td>6269</td>
      <td>3277</td>
      <td>2412</td>
      <td>5588</td>
      <td>6375</td>
      <td>2998</td>
      <td>3460</td>
      <td>5248</td>
      <td>7993</td>
      <td>6069</td>
      <td>6253</td>
      <td>11803</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>1</td>
      <td>100001</td>
      <td>200000</td>
      <td>10857</td>
      <td>5241</td>
      <td>5871</td>
      <td>8727</td>
      <td>6080</td>
      <td>3516</td>
      <td>2983</td>
      <td>5895</td>
      <td>6670</td>
      <td>3174</td>
      <td>3758</td>
      <td>5578</td>
      <td>7089</td>
      <td>6544</td>
      <td>6567</td>
      <td>11449</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>2</td>
      <td>200001</td>
      <td>300000</td>
      <td>10623</td>
      <td>5144</td>
      <td>6129</td>
      <td>8658</td>
      <td>6350</td>
      <td>3575</td>
      <td>2655</td>
      <td>6386</td>
      <td>6617</td>
      <td>3378</td>
      <td>3616</td>
      <td>5325</td>
      <td>6964</td>
      <td>6869</td>
      <td>6536</td>
      <td>11174</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>3</td>
      <td>300001</td>
      <td>400000</td>
      <td>9548</td>
      <td>5117</td>
      <td>6328</td>
      <td>7979</td>
      <td>6483</td>
      <td>4297</td>
      <td>2771</td>
      <td>6834</td>
      <td>6625</td>
      <td>3923</td>
      <td>4047</td>
      <td>5310</td>
      <td>6315</td>
      <td>7049</td>
      <td>6759</td>
      <td>10614</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>4</td>
      <td>400001</td>
      <td>500000</td>
      <td>11515</td>
      <td>5294</td>
      <td>5692</td>
      <td>9068</td>
      <td>6225</td>
      <td>3578</td>
      <td>2638</td>
      <td>5896</td>
      <td>6455</td>
      <td>3027</td>
      <td>3592</td>
      <td>5271</td>
      <td>7375</td>
      <td>6438</td>
      <td>6422</td>
      <td>11513</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>1190</th>
      <td>chr5</td>
      <td>267</td>
      <td>26700001</td>
      <td>26800000</td>
      <td>11432</td>
      <td>5154</td>
      <td>5970</td>
      <td>9120</td>
      <td>6440</td>
      <td>3528</td>
      <td>2620</td>
      <td>5942</td>
      <td>6521</td>
      <td>3051</td>
      <td>3289</td>
      <td>5280</td>
      <td>7283</td>
      <td>6797</td>
      <td>6262</td>
      <td>11310</td>
    </tr>
    <tr>
      <th>1191</th>
      <td>chr5</td>
      <td>268</td>
      <td>26800001</td>
      <td>26900000</td>
      <td>11549</td>
      <td>5299</td>
      <td>5782</td>
      <td>9116</td>
      <td>6113</td>
      <td>3457</td>
      <td>2948</td>
      <td>5881</td>
      <td>6405</td>
      <td>3068</td>
      <td>3254</td>
      <td>5222</td>
      <td>7680</td>
      <td>6575</td>
      <td>5965</td>
      <td>11685</td>
    </tr>
    <tr>
      <th>1192</th>
      <td>chr5</td>
      <td>269</td>
      <td>26900001</td>
      <td>26975502</td>
      <td>8826</td>
      <td>4001</td>
      <td>4609</td>
      <td>6658</td>
      <td>4719</td>
      <td>2639</td>
      <td>2158</td>
      <td>4387</td>
      <td>5163</td>
      <td>2352</td>
      <td>2573</td>
      <td>3979</td>
      <td>5386</td>
      <td>4912</td>
      <td>4726</td>
      <td>8413</td>
    </tr>
    <tr>
      <th>1193</th>
      <td>chr5</td>
      <td>269</td>
      <td>26900001</td>
      <td>26975502</td>
      <td>8826</td>
      <td>4001</td>
      <td>4609</td>
      <td>6658</td>
      <td>4719</td>
      <td>2639</td>
      <td>2158</td>
      <td>4387</td>
      <td>5163</td>
      <td>2352</td>
      <td>2573</td>
      <td>3979</td>
      <td>5386</td>
      <td>4913</td>
      <td>4726</td>
      <td>8413</td>
    </tr>
    <tr>
      <th>1194</th>
      <td>chr5</td>
      <td>269</td>
      <td>26900001</td>
      <td>26975502</td>
      <td>8826</td>
      <td>4001</td>
      <td>4609</td>
      <td>6658</td>
      <td>4719</td>
      <td>2639</td>
      <td>2158</td>
      <td>4387</td>
      <td>5163</td>
      <td>2352</td>
      <td>2573</td>
      <td>3979</td>
      <td>5386</td>
      <td>4914</td>
      <td>4726</td>
      <td>8413</td>
    </tr>
  </tbody>
</table>
<p>1195 rows × 20 columns</p>
</div></div>
</div>
<p>However, a most natural approach is to use <span class="math notranslate nohighlight">\(k=3\)</span>, since the 3-mers represent codons, which encode for amino acids or signal the termination of protein synthesis (stop signals).</p>
<p>The resulting dataframe for arabidopsis 3-mers looks like below.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_3mers_ara</span> <span class="o">=</span> <span class="n">kex</span><span class="o">.</span><span class="n">kmers_by_window_opt</span><span class="p">(</span><span class="n">filepath_ara</span><span class="p">,</span><span class="n">k</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span><span class="n">window_size</span><span class="o">=</span><span class="n">window_size</span><span class="p">,</span><span class="n">output_path</span><span class="o">=</span><span class="n">output_path</span><span class="p">)</span>
<span class="n">df_3mers_tom</span> <span class="o">=</span> <span class="n">kex</span><span class="o">.</span><span class="n">kmers_by_window_opt</span><span class="p">(</span><span class="n">filepath_tom</span><span class="p">,</span><span class="n">k</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span><span class="n">window_size</span><span class="o">=</span><span class="n">window_size</span><span class="p">,</span><span class="n">output_path</span><span class="o">=</span><span class="n">output_path</span><span class="p">)</span>
<span class="n">df_3mers_mz</span> <span class="o">=</span> <span class="n">kex</span><span class="o">.</span><span class="n">kmers_by_window_opt</span><span class="p">(</span><span class="n">filepath_mz</span><span class="p">,</span><span class="n">k</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span><span class="n">window_size</span><span class="o">=</span><span class="n">window_size</span><span class="p">,</span><span class="n">output_path</span><span class="o">=</span><span class="n">output_path</span><span class="p">)</span>
<span class="n">df_3mers_ara</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|████████████████████████████████████████| 1537509/1537509 [00:54&lt;00:00, 28160.74it/s]
100%|██████████████████████████████████████| 10090326/10090326 [05:59&lt;00:00, 28039.88it/s]
100%|██████████████████████████████████████| 25502007/25502007 [15:53&lt;00:00, 26741.24it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chromosome</th>
      <th>window</th>
      <th>start</th>
      <th>end</th>
      <th>AAA</th>
      <th>AAC</th>
      <th>AAG</th>
      <th>AAT</th>
      <th>ACA</th>
      <th>ACC</th>
      <th>...</th>
      <th>TCG</th>
      <th>TCT</th>
      <th>TGA</th>
      <th>TGC</th>
      <th>TGG</th>
      <th>TGT</th>
      <th>TTA</th>
      <th>TTC</th>
      <th>TTG</th>
      <th>TTT</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>0</td>
      <td>1</td>
      <td>100000</td>
      <td>4415</td>
      <td>1966</td>
      <td>2228</td>
      <td>3008</td>
      <td>1926</td>
      <td>1096</td>
      <td>...</td>
      <td>776</td>
      <td>2200</td>
      <td>2016</td>
      <td>1007</td>
      <td>1349</td>
      <td>1881</td>
      <td>2682</td>
      <td>2169</td>
      <td>2422</td>
      <td>4530</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>1</td>
      <td>100001</td>
      <td>200000</td>
      <td>4026</td>
      <td>1914</td>
      <td>2114</td>
      <td>2803</td>
      <td>1849</td>
      <td>1063</td>
      <td>...</td>
      <td>959</td>
      <td>2339</td>
      <td>2102</td>
      <td>957</td>
      <td>1432</td>
      <td>2076</td>
      <td>2324</td>
      <td>2283</td>
      <td>2489</td>
      <td>4353</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>2</td>
      <td>200001</td>
      <td>300000</td>
      <td>3899</td>
      <td>1875</td>
      <td>2138</td>
      <td>2711</td>
      <td>1863</td>
      <td>1066</td>
      <td>...</td>
      <td>932</td>
      <td>2572</td>
      <td>2098</td>
      <td>1085</td>
      <td>1390</td>
      <td>1963</td>
      <td>2255</td>
      <td>2477</td>
      <td>2322</td>
      <td>4120</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>3</td>
      <td>300001</td>
      <td>400000</td>
      <td>3335</td>
      <td>1760</td>
      <td>2117</td>
      <td>2336</td>
      <td>1729</td>
      <td>1110</td>
      <td>...</td>
      <td>833</td>
      <td>2586</td>
      <td>2041</td>
      <td>1313</td>
      <td>1456</td>
      <td>1949</td>
      <td>1998</td>
      <td>2398</td>
      <td>2334</td>
      <td>3884</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>4</td>
      <td>400001</td>
      <td>500000</td>
      <td>4629</td>
      <td>1874</td>
      <td>2085</td>
      <td>2926</td>
      <td>1929</td>
      <td>1075</td>
      <td>...</td>
      <td>885</td>
      <td>2301</td>
      <td>2109</td>
      <td>988</td>
      <td>1406</td>
      <td>1919</td>
      <td>2499</td>
      <td>2241</td>
      <td>2414</td>
      <td>4359</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>1190</th>
      <td>chr5</td>
      <td>267</td>
      <td>26700001</td>
      <td>26800000</td>
      <td>4301</td>
      <td>2000</td>
      <td>2180</td>
      <td>2951</td>
      <td>1818</td>
      <td>1114</td>
      <td>...</td>
      <td>891</td>
      <td>2453</td>
      <td>2014</td>
      <td>1028</td>
      <td>1297</td>
      <td>1922</td>
      <td>2394</td>
      <td>2385</td>
      <td>2340</td>
      <td>4191</td>
    </tr>
    <tr>
      <th>1191</th>
      <td>chr5</td>
      <td>268</td>
      <td>26800001</td>
      <td>26900000</td>
      <td>4423</td>
      <td>1999</td>
      <td>2061</td>
      <td>3066</td>
      <td>1835</td>
      <td>1072</td>
      <td>...</td>
      <td>973</td>
      <td>2304</td>
      <td>1962</td>
      <td>937</td>
      <td>1221</td>
      <td>1845</td>
      <td>2695</td>
      <td>2302</td>
      <td>2196</td>
      <td>4492</td>
    </tr>
    <tr>
      <th>1192</th>
      <td>chr5</td>
      <td>269</td>
      <td>26900001</td>
      <td>26975502</td>
      <td>3366</td>
      <td>1522</td>
      <td>1803</td>
      <td>2135</td>
      <td>1467</td>
      <td>797</td>
      <td>...</td>
      <td>704</td>
      <td>1734</td>
      <td>1585</td>
      <td>738</td>
      <td>960</td>
      <td>1443</td>
      <td>1861</td>
      <td>1717</td>
      <td>1716</td>
      <td>3119</td>
    </tr>
    <tr>
      <th>1193</th>
      <td>chr5</td>
      <td>269</td>
      <td>26900001</td>
      <td>26975502</td>
      <td>3366</td>
      <td>1522</td>
      <td>1803</td>
      <td>2135</td>
      <td>1467</td>
      <td>797</td>
      <td>...</td>
      <td>704</td>
      <td>1734</td>
      <td>1585</td>
      <td>738</td>
      <td>960</td>
      <td>1443</td>
      <td>1861</td>
      <td>1717</td>
      <td>1716</td>
      <td>3119</td>
    </tr>
    <tr>
      <th>1194</th>
      <td>chr5</td>
      <td>269</td>
      <td>26900001</td>
      <td>26975502</td>
      <td>3366</td>
      <td>1522</td>
      <td>1803</td>
      <td>2135</td>
      <td>1467</td>
      <td>797</td>
      <td>...</td>
      <td>704</td>
      <td>1734</td>
      <td>1585</td>
      <td>738</td>
      <td>960</td>
      <td>1443</td>
      <td>1861</td>
      <td>1717</td>
      <td>1716</td>
      <td>3119</td>
    </tr>
  </tbody>
</table>
<p>1195 rows × 68 columns</p>
</div></div>
</div>
</section>
<section id="Read-k-mer-files-if-already-computed">
<h2>Read k-mer files if already computed<a class="headerlink" href="#Read-k-mer-files-if-already-computed" title="Permalink to this headline"></a></h2>
<p>Calculating k-mers can take longer for larger genomes. That’s why you also have the option to load previously saved files, so you don’t have to recalculate the k-mers every time you want to plot some figures or do further analysis.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_2mers_ara</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">+</span><span class="s1">&#39;/CGRW_k2_arabidopsis_genome.csv&#39;</span><span class="p">)</span>
<span class="n">df_2mers_tom</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">+</span><span class="s1">&#39;/CGRW_k2_tomato_genome.csv&#39;</span><span class="p">)</span>
<span class="n">df_2mers_mz</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">+</span><span class="s1">&#39;/CGRW_k2_maize_genome.csv&#39;</span><span class="p">)</span>

<span class="n">df_3mers_ara</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">+</span><span class="s1">&#39;/CGRW_k3_arabidopsis_genome.csv&#39;</span><span class="p">)</span>
<span class="n">df_3mers_tom</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">+</span><span class="s1">&#39;/CGRW_k3_tomato_genome.csv&#39;</span><span class="p">)</span>
<span class="n">df_3mers_mz</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">+</span><span class="s1">&#39;/CGRW_k3_maize_genome.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Group-all-windows-within-each-chromosome">
<h2>Group all windows within each chromosome<a class="headerlink" href="#Group-all-windows-within-each-chromosome" title="Permalink to this headline"></a></h2>
<p>If you don’t want the k-mer frequencies by windows within the chromosomes, you can get the cumulative frequencies for each chromosome as shown in the cell below. For example, the resulting dataframe for arabidopsis would have only 5 rows corresponding to its 5 chromosomes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_2mers_ara_bychr</span> <span class="o">=</span> <span class="n">df_2mers_ara</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;chromosome&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span><span class="n">k</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">df_2mers_ara</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">4</span><span class="p">:]})</span>
<span class="n">df_2mers_tom_bychr</span> <span class="o">=</span> <span class="n">df_2mers_tom</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;chromosome&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span><span class="n">k</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">df_2mers_tom</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">4</span><span class="p">:]})</span>
<span class="n">df_2mers_mz_bychr</span> <span class="o">=</span> <span class="n">df_2mers_mz</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;chromosome&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span><span class="n">k</span><span class="p">:</span><span class="s1">&#39;sum&#39;</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">df_2mers_mz</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">4</span><span class="p">:]})</span>
<span class="n">df_2mers_ara_bychr</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>AA</th>
      <th>AC</th>
      <th>AG</th>
      <th>AT</th>
      <th>CA</th>
      <th>CC</th>
      <th>CG</th>
      <th>CT</th>
      <th>GA</th>
      <th>GC</th>
      <th>GG</th>
      <th>GT</th>
      <th>TA</th>
      <th>TC</th>
      <th>TG</th>
      <th>TT</th>
    </tr>
    <tr>
      <th>chromosome</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>chr1</th>
      <td>3519572</td>
      <td>1588202</td>
      <td>1794811</td>
      <td>2806911</td>
      <td>1923811</td>
      <td>1016038</td>
      <td>697362</td>
      <td>1798090</td>
      <td>1930160</td>
      <td>902322</td>
      <td>1008839</td>
      <td>1579754</td>
      <td>2335983</td>
      <td>1928721</td>
      <td>1920061</td>
      <td>3512205</td>
    </tr>
    <tr>
      <th>chr2</th>
      <td>2291405</td>
      <td>1029105</td>
      <td>1160640</td>
      <td>1834416</td>
      <td>1246754</td>
      <td>672290</td>
      <td>457569</td>
      <td>1166315</td>
      <td>1249456</td>
      <td>584680</td>
      <td>660355</td>
      <td>1026228</td>
      <td>1527955</td>
      <td>1256858</td>
      <td>1242154</td>
      <td>2289303</td>
    </tr>
    <tr>
      <th>chr3</th>
      <td>2697289</td>
      <td>1228414</td>
      <td>1405644</td>
      <td>2153303</td>
      <td>1498561</td>
      <td>804334</td>
      <td>559027</td>
      <td>1396347</td>
      <td>1516501</td>
      <td>714668</td>
      <td>805341</td>
      <td>1226134</td>
      <td>1772296</td>
      <td>1510857</td>
      <td>1492635</td>
      <td>2672182</td>
    </tr>
    <tr>
      <th>chr4</th>
      <td>2143926</td>
      <td>980389</td>
      <td>1109010</td>
      <td>1707158</td>
      <td>1184064</td>
      <td>637226</td>
      <td>439583</td>
      <td>1110444</td>
      <td>1195069</td>
      <td>560350</td>
      <td>630128</td>
      <td>970513</td>
      <td>1417419</td>
      <td>1193357</td>
      <td>1177332</td>
      <td>2125863</td>
    </tr>
    <tr>
      <th>chr5</th>
      <td>3140913</td>
      <td>1410681</td>
      <td>1608073</td>
      <td>2510382</td>
      <td>1712226</td>
      <td>907201</td>
      <td>634608</td>
      <td>1605965</td>
      <td>1741333</td>
      <td>808346</td>
      <td>914520</td>
      <td>1422652</td>
      <td>2075592</td>
      <td>1733770</td>
      <td>1729636</td>
      <td>3160014</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
<section id="Group-all-chromosomes">
<h2>Group all chromosomes<a class="headerlink" href="#Group-all-chromosomes" title="Permalink to this headline"></a></h2>
<p>Furthermore, you can pool the k-mer frequencies of all the chromosomes and have the cumulative counts of the entire sequence.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_2mers_all</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="n">df_2mers_all</span><span class="p">[</span><span class="s1">&#39;arabidopsis&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_2mers_ara_bychr</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="n">df_2mers_all</span><span class="p">[</span><span class="s1">&#39;tomato&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_2mers_tom_bychr</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="n">df_2mers_all</span><span class="p">[</span><span class="s1">&#39;maize&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_2mers_mz_bychr</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="n">df_2mers_all</span><span class="o">.</span><span class="n">T</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>AA</th>
      <th>AC</th>
      <th>AG</th>
      <th>AT</th>
      <th>CA</th>
      <th>CC</th>
      <th>CG</th>
      <th>CT</th>
      <th>GA</th>
      <th>GC</th>
      <th>GG</th>
      <th>GT</th>
      <th>TA</th>
      <th>TC</th>
      <th>TG</th>
      <th>TT</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>arabidopsis</th>
      <td>13793105</td>
      <td>6236791</td>
      <td>7078178</td>
      <td>11012170</td>
      <td>7565416</td>
      <td>4037089</td>
      <td>2788149</td>
      <td>7077161</td>
      <td>7632519</td>
      <td>3570366</td>
      <td>4019183</td>
      <td>6225281</td>
      <td>9129245</td>
      <td>7623563</td>
      <td>7561818</td>
      <td>13759567</td>
    </tr>
    <tr>
      <th>tomato</th>
      <td>85677817</td>
      <td>37700236</td>
      <td>41002572</td>
      <td>75901364</td>
      <td>45946447</td>
      <td>25138712</td>
      <td>11804902</td>
      <td>40969026</td>
      <td>42920385</td>
      <td>18121904</td>
      <td>25183170</td>
      <td>37739813</td>
      <td>65737185</td>
      <td>42898271</td>
      <td>45974544</td>
      <td>85737480</td>
    </tr>
    <tr>
      <th>maize</th>
      <td>161563751</td>
      <td>108257047</td>
      <td>128688710</td>
      <td>136989307</td>
      <td>133796230</td>
      <td>120957040</td>
      <td>87977339</td>
      <td>128781627</td>
      <td>129723428</td>
      <td>112442579</td>
      <td>121098322</td>
      <td>108465666</td>
      <td>110415392</td>
      <td>129855155</td>
      <td>133966157</td>
      <td>161938131</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
<section id="k-mer-figures">
<h2>k-mer figures<a class="headerlink" href="#k-mer-figures" title="Permalink to this headline"></a></h2>
<section id="Plot-k-mers-across-windows">
<h3>Plot k-mers across windows<a class="headerlink" href="#Plot-k-mers-across-windows" title="Permalink to this headline"></a></h3>
<p>With the <code class="docutils literal notranslate"><span class="pre">plot_kmers_across_windows</span></code> function you can plot the frequencies of the k-mers of your choice across the windows of a specified chromosome.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">a</span> <span class="o">=</span> <span class="n">kex</span><span class="o">.</span><span class="n">plot_kmers_across_windows</span><span class="p">(</span><span class="n">df_2mers_ara</span><span class="p">,[</span><span class="s1">&#39;TT&#39;</span><span class="p">,</span><span class="s1">&#39;CC&#39;</span><span class="p">],</span><span class="s1">&#39;chr1&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/kmerExtractorTutorial_22_0.png" src="_images/kmerExtractorTutorial_22_0.png" />
</div>
</div>
<p>You can also use this function to compare, placing the plot of different chromosomes on different axes within the same figure.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span><span class="n">gridspec_kw</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;hspace&#39;</span><span class="p">:</span> <span class="mf">0.3</span><span class="p">})</span>

<span class="n">kex</span><span class="o">.</span><span class="n">plot_kmers_across_windows</span><span class="p">(</span><span class="n">df_3mers_ara</span><span class="p">,[</span><span class="s1">&#39;TTT&#39;</span><span class="p">,</span><span class="s1">&#39;CCC&#39;</span><span class="p">],</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">kex</span><span class="o">.</span><span class="n">plot_kmers_across_windows</span><span class="p">(</span><span class="n">df_3mers_ara</span><span class="p">,[</span><span class="s1">&#39;TTT&#39;</span><span class="p">,</span><span class="s1">&#39;CCC&#39;</span><span class="p">],</span><span class="s1">&#39;chr4&#39;</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/kmerExtractorTutorial_24_0.png" src="_images/kmerExtractorTutorial_24_0.png" />
</div>
</div>
</section>
<section id="Plot-k-mer-frequencies-within-each-chromosome">
<h3>Plot k-mer frequencies within each chromosome<a class="headerlink" href="#Plot-k-mer-frequencies-within-each-chromosome" title="Permalink to this headline"></a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">plot_kmers_freq_within_chromosomes</span></code> function makes a stacked bar graph of the frequency of different k-mers per chromosome. For the same k-mer, each stacked bar represents a different chromosome of the corresponding color.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">b</span> <span class="o">=</span> <span class="n">kex</span><span class="o">.</span><span class="n">plot_kmers_freq_within_chromosomes</span><span class="p">(</span><span class="n">df_2mers_ara</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/kmerExtractorTutorial_27_0.png" src="_images/kmerExtractorTutorial_27_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">c</span> <span class="o">=</span> <span class="n">kex</span><span class="o">.</span><span class="n">plot_kmers_freq_within_chromosomes</span><span class="p">(</span><span class="n">df_3mers_ara</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/kmerExtractorTutorial_28_0.png" src="_images/kmerExtractorTutorial_28_0.png" />
</div>
</div>
</section>
<section id="Plot-FCGR-structure">
<h3>Plot FCGR structure<a class="headerlink" href="#Plot-FCGR-structure" title="Permalink to this headline"></a></h3>
<p>The technique to show the k-mer frequencies as images is known as Frequency Chaos Game Representation (FCGR). It corresponds to a matrix of dimensions <span class="math notranslate nohighlight">\(2^k \times 2^k\)</span>; with the upper left corner equal to the k-mer of C, the upper right corner the k-mer of G, the lower left corner the k-mer of A, and the lower right corner the k-mer of T.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">plot_FCGR</span></code> function plots the FCGR of a sequence as a heatmap that allows to appreciate the relative frequencies of each k-mer. You can plot the FCGR for a specific chromosome and window, the cumulative frequency for an entire chromosome (if no window is specified), or the cumulative frequency for the entire genome (if neither chromosome nor window is specified).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">d</span> <span class="o">=</span> <span class="n">kex</span><span class="o">.</span><span class="n">plot_FCGR</span><span class="p">(</span><span class="n">df_3mers_ara</span><span class="p">,</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span><span class="n">window</span><span class="o">=</span><span class="mi">145</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/kmerExtractorTutorial_31_0.png" src="_images/kmerExtractorTutorial_31_0.png" />
</div>
</div>
<p>You can also adjust the size of the figure and the colormap to your preference as shown below</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">e</span> <span class="o">=</span> <span class="n">kex</span><span class="o">.</span><span class="n">plot_FCGR</span><span class="p">(</span><span class="n">df_2mers_tom</span><span class="p">,</span><span class="s1">&#39;chr6&#39;</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">3</span><span class="p">),</span><span class="n">colormap</span><span class="o">=</span><span class="s1">&#39;Greys&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/kmerExtractorTutorial_33_0.png" src="_images/kmerExtractorTutorial_33_0.png" />
</div>
</div>
<p>Aditionally, is possible compare the FCGR of multiple genomes as shown in the following cell</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>

<span class="n">kex</span><span class="o">.</span><span class="n">plot_FCGR</span><span class="p">(</span><span class="n">df_3mers_ara</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">kex</span><span class="o">.</span><span class="n">plot_FCGR</span><span class="p">(</span><span class="n">df_3mers_tom</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">kex</span><span class="o">.</span><span class="n">plot_FCGR</span><span class="p">(</span><span class="n">df_3mers_mz</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>

<span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Arabidopsis&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Tomato&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Maize&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/kmerExtractorTutorial_35_0.png" src="_images/kmerExtractorTutorial_35_0.png" />
</div>
</div>
</section>
<section id="Plot-CGR">
<h3>Plot CGR<a class="headerlink" href="#Plot-CGR" title="Permalink to this headline"></a></h3>
<p>Finally, although it is not the main objective of the software, a function to generate the CGR image of a sequence is included. In the following example, the FASTA file of the Arabidopsis genome is read and the CGR of the first 100 kb is generated.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">re</span>

<span class="c1"># Read FASTA file with complete genome separated by chromosomes</span>
<span class="n">f_in</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filepath_ara</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
<span class="n">all_file</span> <span class="o">=</span> <span class="n">f_in</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="n">f_in</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<span class="c1"># join lines</span>
<span class="n">seq_all</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">all_file</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">))</span>
<span class="k">del</span> <span class="n">all_file</span>

<span class="c1"># Split sequence by chromosomes</span>
<span class="n">seq_bychr</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;&gt;chr[1-9]*&#39;</span><span class="p">,</span> <span class="n">seq_all</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]</span>
<span class="k">del</span> <span class="n">seq_all</span>

<span class="c1"># plot CGR of the first chromosome, first window (first 100 Kb)</span>
<span class="n">f</span> <span class="o">=</span> <span class="n">kex</span><span class="o">.</span><span class="n">plot_CGR</span><span class="p">(</span><span class="n">seq_bychr</span><span class="p">[</span><span class="mi">0</span><span class="p">][:</span><span class="mi">100_000</span><span class="p">],</span><span class="n">markersize</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|████████████████████████████████████████| 100000/100000 [00:00&lt;00:00, 1886046.78it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/kmerExtractorTutorial_38_1.png" src="_images/kmerExtractorTutorial_38_1.png" />
</div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="kmerExtractor.html" class="btn btn-neutral float-left" title="kmerExtractor module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Camila Riccio-Rengifo.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>